lib/snippets.js

```diff
@@ -0,0 +1,675 @@
+  wordRegexForSnippets (snippets) {
+    const prefixes = {}
+
+    for (let prefix in snippets) {
+      for (let character of prefix) { prefixes[character] = true }
+    }
+
+    const prefixCharacters = Object.keys(prefixes).join('')
+    return new RegExp(`[${_.escapeRegExp(prefixCharacters)}]+`)
+  },
+
+  // Get the best match snippet for the given prefix text.  This will return
+  // the longest match where there is no exact match to the prefix text.
+  snippetForPrefix (snippets, prefix, wordPrefix) {
+    let longestPrefixMatch = null
+
+    for (let snippetPrefix in snippets) {
+      const snippet = snippets[snippetPrefix]
+      if (prefix.endsWith(snippetPrefix) && (wordPrefix.length <= snippetPrefix.length)) {
+        if ((longestPrefixMatch == null) || (snippetPrefix.length > longestPrefixMatch.prefix.length)) {
+          longestPrefixMatch = snippet
+        }
+      }
+    }
+
+    return longestPrefixMatch
+  },
+
+  getSnippets (editor) {
+    return this.parsedSnippetsForScopes(editor.getLastCursor().getScopeDescriptor())
+  },
+
+  snippetToExpandUnderCursor (editor) {
+    if (!editor.getLastSelection().isEmpty()) { return false }
+    const snippets = this.getSnippets(editor)
+    if (_.isEmpty(snippets)) { return false }
+
+    let prefixData = this.getPrefixText(snippets, editor)
+    if (prefixData) {
+      return this.snippetForPrefix(snippets, prefixData.snippetPrefix, prefixData.wordPrefix)
+    }
+  },
+
+  expandSnippetsUnderCursors (editor) {
+    let snippet = this.snippetToExpandUnderCursor(editor)
+    if (!snippet) { return false }
+
+    this.getStore(editor).observeHistory({
+      undo: event => { this.onUndoOrRedo(editor, event, true) },
+      redo: event => { this.onUndoOrRedo(editor, event, false) }
+    })
+
+    this.findOrCreateMarkerLayer(editor)
+    editor.transact(() => {
+      const cursors = editor.getCursors()
+      for (let cursor of cursors) {
+        const cursorPosition = cursor.getBufferPosition()
+        const startPoint = cursorPosition.translate([0, -snippet.prefix.length], [0, 0])
+        cursor.selection.setBufferRange([startPoint, cursorPosition])
+        this.insert(snippet, editor, cursor)
+      }
+    })
+    return true
+  },
+
+  goToNextTabStop (editor) {
+    let nextTabStopVisited = false
+    for (let expansion of this.getExpansions(editor)) {
+      if (expansion != null ? expansion.goToNextTabStop() : undefined) {
+        nextTabStopVisited = true
+      }
+    }
+    return nextTabStopVisited
+  },
+
+  goToPreviousTabStop (editor) {
+    let previousTabStopVisited = false
+    for (let expansion of this.getExpansions(editor)) {
+      if (expansion != null ? expansion.goToPreviousTabStop() : undefined) {
+        previousTabStopVisited = true
+      }
+    }
+    return previousTabStopVisited
+  },
+
+  getStore (editor) {
+    return EditorStore.findOrCreate(editor)
+  },
+
+  // createMarkerLayer (editor) {
```

Comment:
🔥 I'd agree with your commenting out that this is unused code, just remove it 😉.

Reply:
... I actually forgot I did that, I was just seeing if anything broke